File: mixin-padang-garut/src/main/python/garut/functions/dataset/row_count.py
#
# Copyright (c) 2020-2023 Datamixin.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#
import pandas as pd
from typing import Dict, List
from garut.bearer import Bearer
from garut.contexts.context import Context
from garut.instruction import InstructionFactory
from garut.functions.function import FunctionRegistry
from garut.functions.dataset.dataset_function import DatasetFunction


class RowCount(DatasetFunction):

    FUNCTION_NAME = "RowCount"

    def execute(self, context: Context, options: Dict[str, any]) -> any:
        option = self.getDatasetOption(options)
        if isinstance(option, pd.DataFrame):
            return len(option)
        elif isinstance(option, pd.Series):
            return self.createSeries(option, RowCount.length)
        else:
            raise Exception("Unexpected RowCount '{}' option type {}".format(DatasetFunction.DATASET, type(option)))

    def length(value: any) -> int:
        if value is None:
            return "Null value has no length"
        if isinstance(value, Exception):
            raise value
        if isinstance(value, Bearer):
            value = value.getResult()
        return len(value)


parameters = [DatasetFunction.DATASET]
registry: FunctionRegistry = FunctionRegistry.getInstance()
registry.register(RowCount.FUNCTION_NAME, RowCount, parameters)

factory: InstructionFactory = InstructionFactory.getInstance()
factory.registerInstruction(RowCount.FUNCTION_NAME)

File: mixin-padang-webface/src/main/js/padang/functions/dataset/RowCount.ts
/*
 * Copyright (c) 2020-2023 Datamixin.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>. */
import FunctionPlan from "padang/plan/FunctionPlan";
import FunctionPlanRegistry from "padang/plan/FunctionPlanRegistry";

import DatasetFunction from "padang/functions/dataset/DatasetFunction";

export default class RowCount extends DatasetFunction {

    public static FUNCTION_NAME = "RowCount";

    constructor(
        public dataset: string) {
        super(RowCount.FUNCTION_NAME, dataset);
    }

    public static getPlan(): FunctionPlan {
        let plan = new FunctionPlan(
            RowCount.FUNCTION_NAME,
            "Row Count",
            "Row counts",
            "Get number of rows"
        );
        let parameters = plan.getParameterList();
        parameters.add(DatasetFunction.DATASET_PLAN);
        return plan;
    }
}

let registry = FunctionPlanRegistry.getInstance();
registry.registerPlan(RowCount.getPlan());